go/chat/boxer.go

```diff
@@ -1256,23 +1296,39 @@ func (b *Boxer) boxV1(messagePlaintext chat1.MessagePlaintext, key types.CryptKe
 	return boxed, nil
 }
 
-func (b *Boxer) boxV2(messagePlaintext chat1.MessagePlaintext, baseEncryptionKey types.CryptKey,
-	signingKeyPair libkb.NaclSigningKeyPair) (*chat1.MessageBoxed, error) {
+// V3 is just V2 but with exploding messages support. All future versions after
+// V3 will support exploding messages without a version distinction.
+func (b *Boxer) boxV2orV3(messagePlaintext chat1.MessagePlaintext, baseEncryptionKey types.CryptKey,
+	ephemeralSeed *keybase1.TeamEk, signingKeyPair libkb.NaclSigningKeyPair,
+	version chat1.MessageBoxedVersion) (*chat1.MessageBoxed, error) {
 
 	if messagePlaintext.ClientHeader.MerkleRoot == nil {
 		return nil, NewBoxingError("cannot send message without merkle root", false)
 	}
 
-	derivedEncryptionKey, err := libkb.DeriveSymmetricKey(
+	headerEncryptionKey, err := libkb.DeriveSymmetricKey(
 		libkb.NaclSecretBoxKey(baseEncryptionKey.Material()), libkb.EncryptionReasonChatMessage)
 	if err != nil {
 		return nil, err
 	}
 
+	// Regular messages use the same encryption key for the header and for the
+	// body. Exploding messages use a derived ephemeral key for the body.
+	bodyEncryptionKey := headerEncryptionKey
+	if messagePlaintext.IsExploding() {
+		bodyEncryptionKey, err = libkb.DeriveFromSecret(ephemeralSeed.Seed, libkb.DeriveReasonTeamEKExplodingChat)
+		if err != nil {
+			return nil, err
+		}
+		// The MessagePlaintext supplied by the caller has a Lifetime, but we
```

Comment:
etime now not lifetime ;)

Reply:
I'll leave it as is given our design talk this morning.